core/src/unprocessed_packet_batches.rs

```diff
@@ -0,0 +1,1205 @@
+        unprocessed_packet_batches,
+        &shuffled_packet_locators,
+        working_bank,
+    )
+}
+
+// Iterates packets in buffered batches, returns all unprocessed packet's stake,
+// and its location (batch_index plus packet_index within batch)
+fn get_stakes_and_locators(
+    unprocessed_packet_batches: &UnprocessedPacketBatches,
+    packet_sender_info: &mut Option<PacketSenderInfo>,
+) -> (Vec<u64>, Vec<PacketLocator>) {
+    let mut stakes = Vec::<u64>::with_capacity(TOTAL_BUFFERED_PACKETS);
+    let mut locators = Vec::<PacketLocator>::with_capacity(TOTAL_BUFFERED_PACKETS);
+
+    unprocessed_packet_batches.iter().enumerate().for_each(
+        |(batch_index, deserialized_packet_batch)| {
+            let packet_batch = &deserialized_packet_batch.packet_batch;
+            deserialized_packet_batch
+                .unprocessed_packets
+                .keys()
+                .for_each(|packet_index| {
+                    let p = &packet_batch.packets[*packet_index];
+                    stakes.push(p.meta.weight);
+                    locators.push(PacketLocator {
+                        batch_index,
+                        packet_index: *packet_index,
+                    });
+
+                    if let Some(packet_sender_info) = packet_sender_info {
+                        update_packet_sender_info(packet_sender_info, p);
+                    }
+                });
+        },
+    );
+
+    (stakes, locators)
+}
+
+fn weighted_shuffle(
+    stakes: &[u64],
+    locators: &[PacketLocator],
+    packet_sender_info: &mut Option<PacketSenderInfo>,
+) -> Vec<PacketLocator> {
+    let need_to_shuffle_sender_ips = packet_sender_info.is_some();
+    let mut shuffled_packet_senders_ip = Vec::<IpAddr>::new();
+
+    let mut rng = rand::thread_rng();
+    let shuffled_locators: Vec<PacketLocator> = WeightedShuffle::new(stakes)
+        .unwrap()
+        .shuffle(&mut rng)
+        .map(|i| {
+            if need_to_shuffle_sender_ips {
+                let packet_sender_info = packet_sender_info.as_ref().unwrap();
+                shuffled_packet_senders_ip.push(packet_sender_info.packet_senders_ip[i]);
+            }
+            locators[i].clone()
+        })
+        .collect();
+    if let Some(packet_sender_info) = packet_sender_info {
+        packet_sender_info.packet_senders_ip = shuffled_packet_senders_ip;
+    }
+    shuffled_locators
+}
+
+pub fn prioritize_by_fee(
+    unprocessed_packet_batches: &UnprocessedPacketBatches,
+    locators: &Vec<PacketLocator>,
+    bank: Option<Arc<Bank>>,
+) -> Vec<PacketLocator> {
+    let mut fee_buckets = BTreeMap::<u64, Vec<PacketLocator>>::new();
+    for locator in locators {
+        let fee_per_cu = if let Some(fee_per_cu) =
+            compute_fee_per_cu(unprocessed_packet_batches, locator, &bank)
+        {
+            fee_per_cu
+        } else {
+            // if unable to compute fee-per-cu for the packet, put it to the `0` bucket
+            0u64
+        };
```

Comment:
If we're unable to compute the fee-per-cu doesn't that mean the packet was corrupted and we can drop it entirely?